ding/policy/diayn.py

```diff
@@ -0,0 +1,588 @@
+        """
+        loss_dict = {}
+        data = default_preprocess_learn(
+            data,
+            use_priority=self._priority,
+            use_priority_IS_weight=self._cfg.priority_IS_weight,
+            ignore_done=self._cfg.learn.ignore_done,
+            use_nstep=False
+        )
+        if self._cuda:
+            data = to_device(data, self._device)
+
+        self._learn_model.train()
+        self._target_model.train()
+        obs = data.get('obs')
+        next_obs = data.get('next_obs')
+        reward = data.get('reward')
+        done = data.get('done')
+
+        # predict q value
+        q_value = self._learn_model.forward(data, mode='compute_critic')['q_value']
+
+        # predict target value depend self._value_network.
+        if self._value_network:
+            # predict v value
+            v_value = self._learn_model.forward(obs, mode='compute_value_critic')['v_value']
+            with torch.no_grad():
+                next_v_value = self._target_model.forward(next_obs, mode='compute_value_critic')['v_value']
+        else:
+            # target q value. SARSA: first predict next action, then calculate next q value
+            with torch.no_grad():
+                (mu, sigma) = self._learn_model.forward(next_obs, mode='compute_actor')['logit']
+
+                dist = Independent(Normal(mu, sigma), 1)
+                pred = dist.rsample()
+                next_action = torch.tanh(pred)   #这个东西是pi的output
+                y = 1 - next_action.pow(2) + 1e-6
+                next_log_prob = dist.log_prob(pred).unsqueeze(-1)
+                next_log_prob = next_log_prob - torch.log(y).sum(-1, keepdim=True)
+
+                next_data = {'obs': next_obs, 'action': next_action}
+                target_q_value = self._target_model.forward(next_data, mode='compute_critic')['q_value']
+                # the value of a policy according to the maximum entropy objective
+                if self._twin_critic:
+                    # find min one as target q value
+                    target_q_value = torch.min(target_q_value[0],
+                                               target_q_value[1]) - self._alpha * next_log_prob.squeeze(-1)
+                else:
+                    target_q_value = target_q_value - self._alpha * next_log_prob.squeeze(-1)
+        target_value = next_v_value if self._value_network else target_q_value
+
+        # =================
+        # Calculate a new reward function 
+        # =================
+        # get z_one_hot : this is know before learning
+        if self._include_actions:
+            logits = self._learn_model.forward(next_obs, mode='compute_discriminator')['q_discriminator']
+        else:
+            logits = self._learn_model.forward(next_obs, mode='compute_discriminator')['q_discriminator']
+        reward_pl = -1 * torch.nn.CrossEntropyLoss(target=z_one_hot_reduce, input=logits, reduction='none') # z_one_hot_reduce is z_one_hot without one hot (N, C) --> (N)
```

Comment:
z_one_hot_reduce是从哪里传入的呢？我看上面好像没有这个变量。torch.nn.CrossEntropyLoss()里面一般第一项input是logits,第二项target是label/class,参考[https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html](url)

Reply:
- z_one_hot_reduce should come from the collector. This has not been coded up yet. Regarding torch.nn.CrossEntropyLoss(), sure, I swap the order but I guess it will be fine as long as the keys are correct.
- Great! I will add this up 
- The name stem from tensorflow where placeholder is required for tensors. I will change the name latter on to make it more clear. Thx